### ISSUE TITLE
Android kernel does not merge mem regions with different custom comments

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#1861]
The Android kernel is not merging libc's .bss into a subsequent anon
region, while we are:

```
syscall: mprotect addr=0xb6f55000 size=0x00001000 prot=rw-
SYSLOG_WARNING: get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
        os says: 0xb6f53000-0xb6f56000 prot=0x00000003
        cache says: 0xb6f4a000-0xb6f56000 prot=0x00000003

b6f47000-b6f4a000 rw-p 00069000 b3:16 956        /system/lib/libc.so
b6f4a000-b6f53000 rw-p 00000000 00:00 0 
b6f53000-b6f56000 rw-p 00000000 00:00 0          [anon:linker_alloc]
```

We consider .bss to be data, not image (confirmed on Linux too).

```
update_all_memory_areas: adding: 0xb6f44000-0xb6f4a000 prot=3 type=1 share=1
update_all_memory_areas: adding: 0xb6f4a000-0xb6f53000 prot=3 type=2 share=0
update_all_memory_areas: adding: 0xb6f53000-0xb6f54000 prot=3 type=2 share=0
```

Linux kernel does merge, which is why we do:
```
core/unix/memcache.c:            /* we merge adjacent identical-prot image sections: .bss into .data,
```

Could it be that the Android kernel would merge except for that comment string?

Yes, in fact:

```
<get_memory_info mismatch! (can happen if os combines entries in /proc/pid/maps)
        os says: 0xb6eff000-0xb6f02000 prot=0x00000003
        cache says: 0xb6ef6000-0xb6f02000 prot=0x00000003
>

b6e89000-b6eef000 r-xp 00000000 b3:16 956        /system/lib/libc.so
b6eef000-b6ef0000 ---p 00000000 00:00 0 
b6ef0000-b6ef3000 r--p 00066000 b3:16 956        /system/lib/libc.so
b6ef3000-b6ef6000 rw-p 00069000 b3:16 956        /system/lib/libc.so
b6ef6000-b6eff000 rw-p 00000000 00:00 0 
b6eff000-b6f02000 rw-p 00000000 00:00 0          [anon:linker_alloc]
b6f02000-b6f03000 r--p 00000000 00:00 0 
```

But the history is like this:

```
> grep -E 'process_mmap|sysnum=0x000000ac|munmap|mremap' `ls -1td logs/*0|head -1`/h*
process_mmap(0xb6f03000,0x00002000,0x22,rw-,mmap)
process_mmap(0xb6f02000,0x00001000,0x22,rw-,mmap)
process_mmap(0xb6f01000,0x00001000,0x22,rw-,mmap)
post syscall: sysnum=0x000000ac, result=0x00000000 (0)
process_mmap(0xb6f00000,0x00001000,0x22,rw-,mmap)
post syscall: sysnum=0x000000ac, result=0x00000000 (0)
process_mmap(0xb6eff000,0x00001000,0x2,r--,mmap)
process_mmap(0xb6e89000,0x00076000,0x22,---,mmap)
process_mmap(0xb6e89000,0x00065960,0x12,r-x,mmap)
process_mmap(0xb6ef0000,0x0000553c,0x12,rw-,mmap)
process_mmap(0xb6ef6000,0x00009000,0x32,rw-,mmap)
syscall: munmap addr=0xb6eff000 size=0x00001000
process_mmap(0xb6eff000,0x00001000,0x22,rw-,mmap)
post syscall: sysnum=0x000000ac, result=0x00000000 (0)
process_mmap(0xb6e69000,0x00020000,0x1,r--,mmap)
```

The 0xac is SYS_prctl.
Android has a custom prctl to set the comment for a mem region.

Looking at maps after each mmap we see that the extension backward of the
linker_alloc region is at first merged into libc's .bss:

```
process_mmap(0xb6eff000,0x00001000,0x22,rw-,mmap)
0xb6ef0000-0xb6ef6000 /system/lib/libc.so
0xb6ef6000-0xb6f00000 
0xb6f00000-0xb6f02000 [anon:linker_alloc]
```

It's only after that prctl that it gets split from the bss and merged w/
the anon region.

Plan: add whether there is a comment as a prot flag to memcache.  We'll
assume it's rare to have adjacent same-prot regions w/ different comments
and we won't bother to store the comment in memcache.  We'll then monitor
SYS_prctl and do a query and memcache update there.

**Comment by derekbruening** [comment]
I went ahead and used the prctl as a debug-build-only convenient label:

```
shell@flo:/data/local/tmp $ cat /proc/19037/maps                               
501ec000-501ed000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
501ed000-501f3000 rw-p 00000000 00:00 0          [anon:DynamoRIO-internal]
501f3000-501f5000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
501f5000-501f6000 rw-p 00000000 00:00 0          [anon:DynamoRIO-internal]
501f6000-501fd000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
501fd000-50200000 rw-p 00000000 00:00 0          [anon:DynamoRIO-internal]
50200000-50205000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
50205000-5020e000 rw-p 00000000 00:00 0          [anon:DynamoRIO-internal]
5020e000-50215000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
50215000-50216000 r--p 00000000 00:00 0          [anon:DynamoRIO-internal]
50216000-50223000 rw-p 00000000 00:00 0          [anon:DynamoRIO-internal]
50223000-50225000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
50225000-50228000 rwxp 00000000 00:00 0          [anon:DynamoRIO-internal]
50228000-5022d000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
5022d000-5022e000 rwxp 00000000 00:00 0          [anon:DynamoRIO-internal]
5022e000-581ec000 ---p 00000000 00:00 0          [anon:DynamoRIO-internal]
71000000-712c7000 r-xp 00000000 b3:1e 244850     /data/local/tmp/dr/lib32/debug/libdynamorio.so
712c7000-712d0000 ---p 00000000 00:00 0 
712d0000-7132f000 rw-p 002d0000 b3:1e 244850     /data/local/tmp/dr/lib32/debug/libdynamorio.so
7132f000-71352000 rw-p 00000000 00:00 0 
b6ef5000-b6f02000 r-xp 00000000 b3:16 313        /system/bin/linker
b6f02000-b6f03000 ---p 00000000 00:00 0 
b6f03000-b6f05000 rw-p 0000d000 b3:16 313        /system/bin/linker
b6f05000-b6f06000 rw-p 00000000 00:00 0 
b6f06000-b6f07000 ---p 00000000 00:00 0 
b6f07000-b6f08000 r-xp 00000000 b3:1e 228755     /data/local/tmp/hello.android
b6f08000-b6f09000 rw-p 00000000 b3:1e 228755     /data/local/tmp/hello.android
b6f09000-b6f0a000 rw-p 00000000 00:00 0 
b6f0a000-b6f0b000 ---p 00000000 00:00 0 
b6f0b000-b6f0c000 rw-p 00000000 00:00 0 
beed5000-beef6000 rw-p 00000000 00:00 0          [stack]
ffff0000-ffff1000 r-xp 00000000 00:00 0          [vectors]
```
